### PR TITLE
Port int8 and float versions of space_to_batch to TFLM

### PR DESCRIPTION
**Pull request by njeffrie**
Commit 1 copies the TFLite operator into TFLM
Commit 2 implements basic float, int8 and error checking tests along with float and int8 implementations of space_to_batch_nd

This version requires that the flat size of input matches output, since TFLM does not support tensor resizing.

#45693

### CODE REVIEW

**Comment by gbaned** [comment]
@njeffrie  Can you please resolve conflicts? Thanks!

**Review by stephanboner — APPROVED** [review]
Looks good to me!

File: tensorflow/lite/micro/kernels/space_to_batch_nd.cc [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+  TFLITE_DCHECK(context->AllocatePersistentBuffer != nullptr);
+  return context->AllocatePersistentBuffer(context, sizeof(SpaceToBatchParams));
+}
+
+TfLiteStatus Prepare(TfLiteContext* context, TfLiteNode* node) {
+  TF_LITE_ENSURE_EQ(context, NumInputs(node), 3);
+  TF_LITE_ENSURE_EQ(context, NumOutputs(node), 1);
+
+  SpaceToBatchParams* params =
+      static_cast<SpaceToBatchParams*>(node->user_data);
```
Comment (advaitjain) [inline tensorflow/lite/micro/kernels/space_to_batch_nd.cc:48]: nit: move closer to where its first used

File: tensorflow/lite/micro/kernels/space_to_batch_nd.cc [hunk 2]
```diff
@@ -0,0 +1,124 @@
+/* Copyright 2017 The TensorFlow Authors. All Rights Reserved.
```
Comment (advaitjain) [inline tensorflow/lite/micro/kernels/space_to_batch_nd.cc:1]: 2021

File: tensorflow/lite/micro/kernels/space_to_batch_nd_test.cc [hunk 3]
```diff
@@ -0,0 +1,168 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
```
Comment (advaitjain) [inline tensorflow/lite/micro/kernels/space_to_batch_nd_test.cc:1]: 2021

File: tensorflow/lite/micro/kernels/space_to_batch_nd_test.cc [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+#include "tensorflow/lite/c/common.h"
+#include "tensorflow/lite/micro/kernels/kernel_runner.h"
+#include "tensorflow/lite/micro/test_helpers.h"
+#include "tensorflow/lite/micro/testing/micro_test.h"
+
+namespace tflite {
+namespace testing {
+namespace {
+
+const int kBasicInputOutputSize = 16;
```
Comment (advaitjain) [inline tensorflow/lite/micro/kernels/space_to_batch_nd_test.cc:28]: constexpr

**Review by advaitjain — CHANGES_REQUESTED** [review]
small nits only.

Also, the bluepill build is failing.

File: tensorflow/lite/micro/kernels/space_to_batch_nd.cc (hunk 1 above)
Comment (njeffrie) [reply tensorflow/lite/micro/kernels/space_to_batch_nd.cc:48]: Done.

File: tensorflow/lite/micro/kernels/space_to_batch_nd.cc (hunk 2 above)
Comment (njeffrie) [reply tensorflow/lite/micro/kernels/space_to_batch_nd.cc:1]: Done.

File: tensorflow/lite/micro/kernels/space_to_batch_nd_test.cc (hunk 3 above)
Comment (njeffrie) [reply tensorflow/lite/micro/kernels/space_to_batch_nd_test.cc:1]: Done.

File: tensorflow/lite/micro/kernels/space_to_batch_nd_test.cc (hunk 4 above)
Comment (njeffrie) [reply tensorflow/lite/micro/kernels/space_to_batch_nd_test.cc:28]: done.

**Review by advaitjain — APPROVED** [review]

File: tensorflow/lite/kernels/internal/reference/space_to_batch_nd.h [hunk 5]
```diff
@@ -41,18 +41,10 @@ inline void SpaceToBatchND(const SpaceToBatchParams& params,
...
-    new_shape.SetDim(1, shape.Dims(1));
-    new_shape.SetDim(3, shape.Dims(2));
-    return new_shape;
-  };
-  const RuntimeShape input1_shape = extend_shape(unextended_input1_shape);
-  const RuntimeShape output_shape = extend_shape(unextended_output_shape);
+  const RuntimeShape input1_shape =
+      RuntimeShape::ExtendedShape(4, unextended_input1_shape);
+  const RuntimeShape output_shape =
+      RuntimeShape::ExtendedShape(4, unextended_output_shape);
```
Comment (advaitjain) [inline tensorflow/lite/kernels/internal/reference/space_to_batch_nd.h:47]: is it still worth adding a check for DimensionsCount ==4?

```cc
const RuntimeSHare input1_shape = (unextended_input1_shape.DimensionsCount() == 4) ? unextended_input1_shape : ExtendSHape();
```

**Review by advaitjain — COMMENTED** [review]
should we be safe and keep the existing behavior to avoid unnecessary creation of a RuntimeShape?

File: tensorflow/lite/kernels/internal/reference/space_to_batch_nd.h (hunk 5 above)
Comment (njeffrie) [reply tensorflow/lite/kernels/internal/reference/space_to_batch_nd.h]: Done.

**Review by advaitjain — APPROVED** [review]

File: tensorflow/lite/kernels/internal/reference/space_to_batch_nd.h [hunk 6]
```diff
@@ -18,10 +18,9 @@ limitations under the License.
 #include <cmath>
 
 #include "ruy/profiler/instrumentation.h"  // from @ruy
-#include "tensorflow/lite/kernels/internal/types.h"
+#include "tensorflow/lite/kernels/internal/common.h"
```
Comment (advaitjain) [inline tensorflow/lite/kernels/internal/reference/space_to_batch_nd.h:21]: I'm a bit confused by what this commit is fixing. types.h is ok to be included for TFLM.

In fact, the params are declared in that header:
https://github.com/tensorflow/tensorflow/blob/ef214a5a9d217aedcfcff25027d7eab488385e29/tensorflow/lite/kernels/internal/types.h#L1096-L1099

**Review by advaitjain — COMMENTED** [review]
not sure that changing types.h to common.h will fix anything. Include what you use would actually mean that we should be directly including types.h

**Review by advaitjain — APPROVED** [review]

**Review by advaitjain — APPROVED** [review]

File: tensorflow/lite/kernels/internal/reference/space_to_batch_nd.h (hunk 6 above)
Comment (advaitjain) [reply tensorflow/lite/kernels/internal/reference/space_to_batch_nd.h]: underlying issue what elsewhere, and fixed with [`0ec1f46` (#46714)](https://github.com/tensorflow/tensorflow/pull/46714/commits/0ec1f462f82afe9c0e0d5a95958deb71451bc831)

**Review by advaitjain — APPROVED** [review]

**Review by advaitjain — APPROVED** [review]

**Pull request merged**